x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/space_aware_privilege_section/privilege_space_form.tsx

```diff
@@ -474,32 +501,39 @@ export class PrivilegeSpaceForm extends Component<Props, State> {
   };
 
   private onFeaturePrivilegesChange = (featureId: string, privileges: string[]) => {
-    const role = copyRole(this.state.role);
-    const form = role.kibana[this.state.privilegeIndex];
-
-    if (privileges.length === 0) {
-      delete form.feature[featureId];
-    } else {
-      form.feature[featureId] = [...privileges];
-    }
-
-    this.setState({
-      role,
-      privilegeCalculator: new PrivilegeFormCalculator(this.props.kibanaPrivileges, role),
-    });
+    this.setRole(privileges, featureId);
   };
 
   private onChangeAllFeaturePrivileges = (privileges: string[]) => {
+    this.setRole(privileges);
+  };
+
+  private setRole(privileges: string[], featureId?: string) {
     const role = copyRole(this.state.role);
     const entry = role.kibana[this.state.privilegeIndex];
 
     if (privileges.length === 0) {
-      entry.feature = {};
+      if (featureId) {
+        delete entry.feature[featureId];
+      } else {
+        entry.feature = {};
+      }
     } else {
-      this.props.kibanaPrivileges.getSecuredFeatures().forEach((feature) => {
-        const nextFeaturePrivilege = feature
-          .getPrimaryFeaturePrivileges()
-          .find((pfp) => privileges.includes(pfp.id));
+      let securedFeaturesToSet = this.props.kibanaPrivileges.getSecuredFeatures();
+      if (featureId) {
+        const findSecuredFeature = securedFeaturesToSet.find((sf) => sf.id === featureId);
+        securedFeaturesToSet = findSecuredFeature ? [findSecuredFeature] : securedFeaturesToSet;
+      }
+      securedFeaturesToSet.forEach((feature) => {
+        const nextFeaturePrivilege = feature.getPrimaryFeaturePrivileges().find((pfp) => {
+          if (
+            !pfp.disabled ||
+            (pfp.requireAllSpaces && this.state.selectedSpaceIds.includes(ALL_SPACES_ID))
+          ) {
+            return privileges.includes(pfp.id);
+          }
```

Comment:
I haven't tested this but reading the logic, I think this might be a more accurate conditional:

```suggestion
          if (
            !pfp.disabled &&
            (!pfp.requireAllSpaces || this.state.selectedSpaceIds.includes(ALL_SPACES_ID))
          ) {
            return privileges.includes(pfp.id);
          }
```

WDYT?